File: frontend/src/api/issues.ts
// src/api/issues.ts
import { AxiosInstance } from "axios";
import { components } from "@/api/models/models";

export type IssueCreateRequest = components["schemas"]["IssueCreateRequest"];
export type IssueReadResponse = components["schemas"]["IssueReadResponse"];
export type IssueUpdateRequest = components["schemas"]["IssueUpdateRequest"];
export type IssueUpdateResponse = components["schemas"]["IssueUpdateResponse"];
export type IssueDeleteResponse = components["schemas"]["IssueDeleteResponse"];
export type IssueListResponse = components["schemas"]["IssueListResponse"];
export type StanceReadResponse = components["schemas"]["StanceReadResponse"];

/**
 * Create a new issue (admin only).
 */
export async function createIssue(
  api: AxiosInstance,
  payload: IssueCreateRequest
): Promise<IssueReadResponse> {
  const res = await api.post<IssueReadResponse>("/issues", payload);
  return res.data;
}

/**
 * Fetch a single issue by ID.
 */
export async function getIssue(
  api: AxiosInstance,
  issueId: number
): Promise<IssueReadResponse> {
  const res = await api.get<IssueReadResponse>(`/issues/${issueId}`);
  return res.data;
}

/**
 * Update an issue (admin only).
 */
export async function updateIssue(
  api: AxiosInstance,
  issueId: number,
  payload: IssueUpdateRequest
): Promise<IssueUpdateResponse> {
  const res = await api.put<IssueUpdateResponse>(`/issues/${issueId}`, payload);
  return res.data;
}

/**
 * Delete an issue (admin only).
 */
export async function deleteIssue(
  api: AxiosInstance,
  issueId: number
): Promise<IssueDeleteResponse> {
  const res = await api.delete<IssueDeleteResponse>(`/issues/${issueId}`);
  return res.data;
}

/**
 * List all issues.
 */
export async function listIssues(api: AxiosInstance): Promise<IssueListResponse> {
  const res = await api.get<IssueListResponse>("/issues");
  return res.data;
}

/**
 * Fetch the current user's stance for a specific issue.
 */
export async function getMyStanceForIssue(
  api: AxiosInstance,
  issueId: number
): Promise<StanceReadResponse | null> {
  const res = await api.get<StanceReadResponse | null>(`/issues/${issueId}/stances/me`);
  return res.data;
}

File: backend/app/routers/issues.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from app.dependencies import get_db, get_is_admin, get_current_user
from app.database.issue import create_issue, read_issue, update_issue, delete_issue, get_all_issues
from app.database.stance import get_user_stance_by_issue
from app.routers.models.issues import IssueCreateRequest, IssueReadResponse, IssueUpdateRequest, IssueUpdateResponse, IssueDeleteResponse, IssueListResponse
from app.routers.models.stances import StanceReadResponse
import logging
from typing import Optional

router = APIRouter(tags=["issues"])

@router.post("/issues", response_model=IssueReadResponse)
def create_issue_endpoint(request: IssueCreateRequest, db: Session = Depends(get_db), is_admin: bool = Depends(get_is_admin)) -> IssueReadResponse:
    if not is_admin:
        raise HTTPException(status_code=403, detail="Admin privileges required")
    issue = create_issue(db, title=request.title, description=request.description)
    if not issue:
        raise HTTPException(status_code=400, detail="Failed to create issue")
    return IssueReadResponse(
        id=issue.id,
        title=issue.title,
        description=issue.description,
        created_at=issue.created_at.isoformat(),
        updated_at=issue.updated_at.isoformat()
    )

@router.get("/issues/{issue_id}", response_model=IssueReadResponse)
def get_issue_endpoint(issue_id: int, db: Session = Depends(get_db)) -> IssueReadResponse:
    issue = read_issue(db, issue_id=issue_id)
    if not issue:
        raise HTTPException(status_code=404, detail="Issue not found")
    return IssueReadResponse(
        id=issue.id,
        title=issue.title,
        description=issue.description,
        created_at=issue.created_at.isoformat(),
        updated_at=issue.updated_at.isoformat()
    )

@router.put("/issues/{issue_id}", response_model=IssueUpdateResponse)
def update_issue_endpoint(issue_id: int, request: IssueUpdateRequest, db: Session = Depends(get_db), is_admin: bool = Depends(get_is_admin)) -> IssueUpdateResponse:
    if not is_admin:
        raise HTTPException(status_code=403, detail="Admin privileges required")
    issue = update_issue(db, issue_id=issue_id, **request.dict(exclude_unset=True))
    if not issue:
        raise HTTPException(status_code=400, detail="Failed to update issue")
    return IssueUpdateResponse(
        id=issue.id,
        title=issue.title,
        description=issue.description,
        created_at=issue.created_at.isoformat(),
        updated_at=issue.updated_at.isoformat()
    )

@router.delete("/issues/{issue_id}", response_model=IssueDeleteResponse)
def delete_issue_endpoint(issue_id: int, db: Session = Depends(get_db), is_admin: bool = Depends(get_is_admin)) -> IssueDeleteResponse:
    if not is_admin:
        raise HTTPException(status_code=403, detail="Admin privileges required")
    success = delete_issue(db, issue_id=issue_id)
    if not success:
        raise HTTPException(status_code=400, detail="Failed to delete issue")
    return IssueDeleteResponse(success=True)

@router.get("/issues", response_model=IssueListResponse)
def get_all_issues_endpoint(db: Session = Depends(get_db)):
    issues = get_all_issues(db)
    return IssueListResponse(
        issues=[
            IssueReadResponse(
                id=issue.id,
                title=issue.title,
                description=issue.description,
                created_at=issue.created_at.isoformat(),
                updated_at=issue.updated_at.isoformat()
            ) for issue in issues
        ]
    )

@router.get("/issues/{issue_id}/stances/me", response_model=Optional[StanceReadResponse])
def get_my_stance_for_issue(issue_id: int, db: Session = Depends(get_db), user_id: int = Depends(get_current_user)) -> Optional[StanceReadResponse]:
    stance = get_user_stance_by_issue(db, issue_id=issue_id, user_id=user_id)
    if not stance:
        return None
    return StanceReadResponse(
        id=stance.id,
        user_id=stance.user_id,
        event_id=stance.event_id,
        issue_id=stance.issue_id,
        headline=stance.headline,
        content_json=stance.content_json
    )
